File: 백준/Gold/17281. ⚾/⚾.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;

    static int N;
    static int[][] map;
    static int[] selected;
    static boolean[] visited;
    static int max;

    public static void main(String[] args) throws IOException {

        input();

        //1~9까지 브루트포스로 selected에 넣기. 단. 1번은 반드시 4번째 !!
        selected = new int[10];
        visited = new boolean[10];
        selected[4] = 1;
        max = 0;
        dfs(1, 2);

        System.out.println(max);




    }

    public static void dfs(int cnt, int idx){
        if(cnt==10){
            //selected에 있는 애들 가지고 최대점수 계산
            baseball();

            return;

        }

        for(int i=2; i<=9; i++){
            if(!visited[i]){
                visited[i] = true;
                selected[cnt] = i;
                if (cnt + 1 == 4) {
                    dfs(cnt + 2, i + 1);
                } else {
                    dfs(cnt + 1, i + 1);
                }
                visited[i] = false;
            }
        }

    }

    public static void baseball(){
        int inning = 0;
        int idx = 1; //시작 선수 번호



        Queue<int[]> queue = new LinkedList<>();//현재 1,2,3루에 있는 선수 번호
        //selected의 순서대로 이닝을 돈다

        int score = 0;
        int out = 0;
        while(inning<N) {



            int result = map[inning][selected[idx]-1]; //현재 선수의 결과

            if (result >= 1) {

                score += run(queue, result, selected[idx]);
                idx++;

                if (idx == 10) idx = 1;
            } else {
                out++;
                idx++;
                if (idx == 10) idx = 1;
                if (out == 3) {
                    out = 0;
                    inning++;
                    queue.clear();
                    if (inning == N) {
                        max = Math.max(max, score);
                        return;
                    }
                }
            }
        }



    }

    public static int run(Queue<int[]> queue, int num, int batter){

        int score =0;
        int size = queue.size();
        for(int i=0; i<size; i++){
            int[] runner = queue.poll();
            runner[1] += num;
            if(runner[1]>=4){
                score++;
            }else{
                queue.add(runner);
            }
        }

        if(num==4) score++;
        else queue.add(new int[]{batter, num});
        return score;
    }


    public static void input() throws IOException {
        N = Integer.parseInt(br.readLine());
        map = new int[N][9];

        for(int i=0; i<N; i++){
            st = new StringTokenizer(br.readLine());
            for(int j=0; j<9; j++){
                map[i][j] = Integer.parseInt(st.nextToken());
            }
        }

    }
}

File: 백준/Gold/17281. ⚾/⚾.py
from collections import deque
import sys

input = sys.stdin.readline
n = int(input())
innings = [list(map(int, input().split())) for _ in range(n)]

selected = [0] * 9
selected[3] = 0
visited = [0] * 9


def select_player(cnt):
    global answer
    if cnt == 9:
        count_score()
        return

    for i in range(1, 9):
        if not visited[i]:
            visited[i] = 1
            selected[cnt] = i
            if cnt + 1 == 3:
                select_player(cnt + 2)
            else:
                select_player(cnt + 1)
            visited[i] = 0


def count_score():
    idx = 0
    score = 0
    global answer
    for inn in innings:
        out = 0
        # ru = [0] * 3
        ru = deque([])
        while 1:
            now = inn[selected[idx]]
            if now == 0:
                out += 1
                if out >= 3:
                    idx += 1
                    idx %= 9
                    break
            elif now <= 4:
                size = len(ru)
                for k in range(size):
                    runner = ru.popleft()
                    runner += now
                    if runner>=4:
                        score+= 1
                    else:
                        ru.append(runner)
                if now == 4:
                    score+=1
                else:
                    ru.append(now)
            #     for k in range(3 - now, 3):
            #         score += ru[k]
            #         ru[k] = 0
            #     for k in range(2, now - 1, -1):
            #         ru[k] = ru[k - now]
            #         ru[k - now] = 0
            #     ru[now - 1] = 1
            #
            # else:
            #     score += sum(ru) + 1
            #     ru = [0] * 3
            idx += 1
            idx %= 9

    answer = max(answer, score)



answer = 0
select_player(0)
print(answer)
